src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,174 @@
+        } else {
+            nodeToRemove.previous.next = nodeToRemove.next;
+            nodeToRemove.next.previous = nodeToRemove.previous;
+        }
+        nodeToRemove.next = nodeToRemove.previous = null;
+    }
+
+    public Node<T> findNode(int index) {
+        Node<T> headHandler = head;
+        Node<T> tailHandler = tail;
+        if (index < size / 2) {
+            for (int i = 0; i < index && headHandler.next != null; i++) {
+                headHandler = headHandler.next;
+            }
+            return headHandler;
+        } else {
+            for (int i = size - 1; i > index && tailHandler.previous != null; i--) {
+                tailHandler = tailHandler.previous;
+            }
+            return tailHandler;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> nodeToAdd = new Node<>(value);
+        if (head == null) {
+            nodeToAdd.previous = null;
+            nodeToAdd.next = null;
+            head = nodeToAdd;
+            tail = nodeToAdd;
+        } else {
+            nodeToAdd.next = null;
+            nodeToAdd.previous = tail;
+            tail.next = nodeToAdd;
+            tail = nodeToAdd;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkBounds(index);
+        nodeInsertion(findNode(index).previous, new Node<>(value), findNode(index));
 
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkBounds(index);
+        return findNode(index).storedItem;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkBounds(index);
+        Node<T> nodeInPlace = findNode(index);
+        T elementToReplace = nodeInPlace.storedItem;
+        nodeInPlace.storedItem = value;
+        return elementToReplace;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkBounds(index);
+        Node<T> nodeToRemove = findNode(index);
+        removeOnIndex(index, nodeToRemove);
+        --size;
+        return nodeToRemove.storedItem;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> holder = head;
+        int i = 0;
+        while (i < size && holder.next != null && !(holder.storedItem == t
+                || (holder.storedItem != null && holder.storedItem.equals(t)))) {
+            holder = holder.next;
+            i++;
+        }
+        if (holder.storedItem == t || (holder.storedItem != null && holder.storedItem.equals(t))) {
+            removeOnIndex(i, holder);
+            size--;
+            return true;
+        }
```

Comment:
```suggestion
        while (holder.next != null) {
            if (holder.storedItem == t || (holder.storedItem != null && holder.storedItem.equals(t))) {
                   unlink(holder);
                   size--;
                   return true;
            }
            holder = holder.next;
        }
```

Reply:
but I'm already using it, yet in old signature, on line 160. What does this method has with refactoring remove(T t)? The best option that I can see to refactor it is to replace while cycle with for cycle

Reply:
done